man/man1/kmquery.1
.ie \(.g \{\
.\" ... groff (hack for khelpcenter, man2html, etc.)
.TH KMQUERY 1 "" "Performance Co-Pilot"
\}
.el \{\
.if \nX=0 .ds x} KMQUERY 1 "" "Performance Co-Pilot"
.if \nX=1 .ds x} KMQUERY 1 "Performance Co-Pilot"
.if \nX=2 .ds x} KMQUERY 1 "" "\&"
.if \nX=3 .ds x} KMQUERY "" "" "\&"
.TH \*(x}
.rr X
\}
.SH NAME
\f3kmconfirm\f1, \f3kmmessage\f1, \f3kmquery\f1 \- general purpose dialog box
.SH SYNOPSIS
\f3kmconfirm\f1
[\f3\-c\f1]
[\f3\-b\f1 \f2button-name\f1]
[\f3\-B\f1 \f2default-button-name\f1]
[\f3\-t\f1 \f2string\f1]
[\f3\-file\f1 \f2filename\f1]
[\f3\-icon\f1 \f2icontype\f1]
[\f3\-font\f1 \f2font\f1]
[\f3\-header\f1 \f2titlebar-string\f1]
[\f3\-useslider\f1]
[\f3\-noslider\f1]
[\f3\-noframe\f1]
[\f3\-exclusive\f1]
.br
.PP
\f3kmmessage\f1
[\f3\-buttons\f1 \f2label1\f1[:\f2value1\f1][,\f2label2\f1[:\f2value2\f1][,...]]]
[\f3\-center\f1]
[\f3\-nearmouse\f1]
[\f3\-default\f1 \f2button\f1]
[\f3\-file\f1 \f2filename\f1]
[\f3\-print\f1]
[\f3\-timeout\f1 \f2sec\f1]
.I message...
.br
.PP
\f3kmquery
[\f3\-input\f1]
[\f3all above options...\f1]
[\f2message...\f1]
.SH DESCRIPTION
.B kmquery
provides a command-line-option compatible implementation of the
.B xconfirm
and
.B xmessage
tools, using a look-and-feel that is consistent with
.BR kmchart .
Several extensions to the functionality of the original tools have been made,
in order to improve their specific utility for
.BR kmchart ,
but wherever possible the original semantics remain.
.PP
.B kmconfirm
displays a line of text for each
.I \-t
argument specified (or a file when the
.I \-file
argument is used),
and a button for each
.I \-b
argument specified.
When one of the buttons is pressed, the label of that button is written to
.B kmquery's
standard output.
This provides a means of communication/feedback from within shell
scripts and a means to display useful information to a user from
an application.
.PP
.B kmmessage
displays a window containing a message from the command line, a file,
or standard input.
It additionally allows buttons to be associated with an exit status,
and only optionally will write the label of the button to standard output.
.PP
.B kmquery
extends the above tools to additionally support limited user input,
as free form text.
In this
.B \-input
mode, any text entered will be output when the default button is pressed.
A default text can be entered using the same mechanisms as the other tools.
.PP
Command line options are available to specify font style, frame style,
modality and one of several different icons to be presented for tailored
visual feedback to the user.
.TP 5
.B \-c \f1or \f3\-center\f1
Center the window on the display.
.TP
.B \-nearmouse
Pop up the window near the mouse cursor.
.TP
.B \-b \f2button-name\f1
Displays a button with the label
.IR button-name .
If
.I button-name
is the empty string, the button in that position is not displayed.
If no
.B \-b
arguments are present, the default is a button with the label Continue.
The exit status associated with
.I button-name
is zero.
.TP
.B \-B \f2button-name\f1
Displays a button with the label
.I button-name
and specifies it as the button to be activated when enter is pressed.
The exit status associated with
.I button-name
is zero.
.TP
.B \-buttons \f2button,button,.\|.\|.\f1
This option will create one button for each comma-separated \f2button\f1
argument.
Each \f2button\f1 consists of a label optionally followed by a colon
and an exit value.
The exit value will be returned if that button is selected.
The default exit value is 100 plus the button number.
Buttons are numbered from the left starting with one.
.TP
.B \-default \fIlabel\fP
Defines the button with a matching \fIlabel\fP to be the default.
If not specified there is no default.
The corresponding resource is \fBdefaultButton\fP.
Pressing Return anywhere in the \fIxmessage\fP window will activate
the default button.
The default button has a wider border than the others.
.TP
.B \-t \f2message\f1
Displays message.
Any number of strings can be listed on the command line
(each must be preceded with the
.B \-t
option).
.TP
.B \-file \f2filename\f1
Displays the file
.I filename.
All
.B \-t
options will be ignored.
A \f2filename\f1 of `\f2\-\f1' reads from standard input.
.TP
.B \-icon \f2icontype\f1
Displays the icon
.I icontype
where icontype is one of:
.IR info ,
.IR error ,
.IR question ,
.IR warning ,
.IR critical .
.I action
is also accepted as a synonym for
.I error
for backward compatibility.
.BR kmquery
introduces the additional
.I archive
and
.I host
icon types as well as the original 
.BR xconfirm
types listed earlier.
.TP
.B \-font \f2fontname\f1
Use fontname as the font.
This option is only available when using the X Window System.
.TP
.B \-header \f2string\f1
Use string as the window title.
.TP
.B \-print
This causes the program to write the label of the button pressed to
standard output.
It is the default behaviour for
.B kmconfirm
and
.BR kmquery .
.TP
.B \-noprint
This causes the program to not write the label of the button pressed to
standard output.
It is the default behaviour for
.BR kmmessage .
.TP
.B \-geometry \f2geometry-string\f1
This provides xconfirm with an X-compatible geometry string specification.
This option is only available when using the X Window System.
.TP
.B \-useslider
When displaying a file, always use a slider instead of determining
automatically whether a slider is necessary.
.TP
.B \-noslider
Do not create a slider, and clip text to the window size, instead of
determining automatically whether a slider is necessary..
.TP
.B \-noframe
Do not display a frame around the contents.
.TP
.B \-exclusive
Grab the keyboard/pointer and do not allow further
input until a button is pressed.
.TP
.B \-timeout \f2secs\f1
Exit with status 0 after \fIsecs\fP seconds if the user has not
clicked on a button yet.
The corresponding resource is \fBtimeout\fP.
.SH EXAMPLES
The following shell script will display a window with an information icon,
asking the user a yes or no question with "Yes" as the default.
.PP
.nf
 #! /bin/sh
 case `kmquery \-t "Really power down?" \-b No \-B Yes \-icon question
 in
   Yes) shutdown;;
   No) ;;
 esac
.fi
.PP
A second example, which prompts for a hostname then starts a
terminal with an ssh session connected to the requested host.
.PP
.nf
 #! /bin/sh
 host=`kmquery \-input \-icon host \-b Cancel \-B OK \\
               \-header "Remote Terminal \- Secure Shell"
 [ "$host" = "Cancel" \-o \-z "$host" ] && exit
 gnome-terminal \-e "ssh $host"
.fi
.SH ENVIRONMENT
.B kmquery
is an excellent choice of utility for the "PCP_XCONFIRM_PROG"
Performance Co-Pilot configuration parameter (refer to
.BR pcp.conf (4)
for details).
.PP
Note that PCP_XCONFIRM_PROG will be automatically set to
.B kmquery
inside tools like
.BR kmchart ,
unless PCP_XCONFIRM_PROG is already set in the environment.
.SH "EXIT STATUS"
If it detects an error,
.B kmquery
always returns 1, so this value should not be associated with a button.
Unless \f2\-button\f1 option has not been used, the return code will be
zero on success.
.SH "SEE ALSO"
.BR kmchart (1),
.BR xconfirm (1),
.BR xmessage (1),
.BR pcp.conf (4).


CHANGELOG
pcp-2.1.1 (released 7 Dec 1999)
    - initial pcp-2.1.1 open source release

pcp-2.1.2 (released with 2.1.4)
    - released with SGI ACE 1.0
    - fixed date field in LSM
    - install pmiestats.h for pmieconf
    - renamed /var/pcp/config/pmafm/pcp-col to
      /var/pcp/config/pmafm/pcp
    - added network.{ip,icmp,tcp,udp} metrics
      extracted from /proc/net/snmp
    - fixed shell syntax error in pmlogger_daily(1)
      (changed test -l to test -f)
    - fixed problem with definition of __psint_t on 64bit machines
    - fixed timezone problem with daylight saving

pcp-2.1.3 (released with 2.1.4)
    - released with ISE 1.0 (MR 1 Mar 2000)
    - added hinv.map.scsi metric as a map of SCSI devices (if any)
    - added disk.dev.{read,write,total}_bytes
    - filter devpts from filesys.* metrics to avoid FPE errors
    - added support for new style Apache configuration file and
      "silent" install to weblog scripts

pcp-2.1.4 (released 15 Feb 2000)
    - fix for segfault src/pmdas/linux/pmda.c on linuxppc, found by
      Dale F. Brantly <dale.brantly@sgi.com>
    - fix build error in src/pmclient, was trying to use pmgenmap
      without /etc/pcp.env, also found by Dale.
    - setlinebuf(stdout) for all clients that loop
      (fixes a buffering problem on alpha-linux).
    - fixed some missing "{" in conditional variable assignments
      in src/pmie/etc_init.d_pmie (found by lstep@free.fr)
    - changed use of pmgenmap in build for pmclient and
      pmkstat to be consistent (found by lstep@free.fr)
    - don't assume clock tick rate is 100/second. Use the CLK_TCK macro
      instead. Rearrange arithmetic in src/pmdas/linux/pmda.c where we
      divide by CLK_TCK to avoid truncation in conversions from jiffies
      to milliseconds.
    - return "no values available" rather than PM_ERR_APPVERSION for
      the metrics disk.{all,dev}.{read,write,total}_bytes since these
      are not available on systems without the "sard disk patch".  This
      is now consistent with requests for disk.partitions.* metrics.

pcp-2.1.5 (released 12 April 2000)
    - released with SGI ACE 1.2 (MR 19 Apr 200)
    - fix bug #786743 filesys metrics are wrong
    - fix potential segfault bug in timezone handling
    - tolerate SUSE's location of magic file different to Redhat's
    - tolerate no chkconfig on SUSE (use %postinstall create rc symlinks)

pcp-2.1.6 (released 3 May 2000)
    - released with ISE 1.2 (MR 25 May 2000)
    - released with ISEMAIL 1.0 (MR 24 May 2000)
    - this is the version shipped with SGI ProPack1.3
    - change use of __clone to pthread_create for portability
    - add support for raid disk stats (previously ignored)
    - fix for bug #789425 pmie suspect behaviour with "delta" keyword
    - add support for devfs style scsi disk names
    - fixed a problem where nfs stats were always showing NFSv3 stats
      rather than NFSv2 stats. We are not (yet) supporting NFSv3 stats.

pcp-2.1.7-1 (internal release for testing)
    - merged changes from IRIX for multiple namespace support in
      pmdaproc.sh
    - parameterized path to pmcd.conf and pmcd.options to avoid
      having to move these files in IRIX. Use PCP_PMCDCONF_PATH and
      PCP_PMCDOPTIONS_PATH respectfully.
    - add new error define PM_ERR_LOGFILE (reconciled from IRIX)
    - conditional pragma for pmGetConfig (not used in linux)
    - use if defined(HAVE_OBJECT_STYLE) to guard __pmCheckObjectStyle
      since it's not used in linux (reconciled with irix code)
    - correct logic used to scan for pmlogger pid in pmlogger_check
      in case where pmlogger exits prematurely, part of #789025
    - added -Wall to default CFLAGS and LDFLAGS
    - added xfs metrics (needs kernel support else no values available)
      extracted from /proc/fs/xfs/stat
    - added nfs (version 3) metrics, same names as on irix, bug #789669
    - fix instance domain for kernel.percpu.interrupts, bug #790372

pcp-2.1.7-2 (released 15 June 2000)
    - bumped build number ..
    - fixed annoying warning messages about directory not empty
      when doing an upgrade

pcp-2.1.8-2 (released 30 June 2000)
    - fix for bug #793871 pmlogger_check fails after redhat upgrade
      (because PCP entries in /usr/share/magic were clobbered)
    - also install /var/pcp/pmdas/linux/pmdalinux (as a non-DSO
      agent for debugging and profiling purposes).
    - added pmda.uname (uname -a) and pmda.version (linux pmda version)
      metrics. The pmda.uname metric is needed by the "pcp" command.
    - fix for #789025 fix to ensure rpm --verify succeeds immediately
      after an install, and other errors in pmlogger_check
    - released with ACE 1.3 (MR 19 Jul 2000)

pcp-2.1.9-6 (released 2 Aug 2000 for propack1.4 - alpha, not final)
    - install /usr/share/pcp/lib/rc-proc.sh containing common
      shell functions for use by rc scripts - these functions
      are tolerant of the chkconfig command missing (as in SUSE).
    - update all rc scripts and {pmlogger,pmie}_{check,daily} scripts
      to use the new rc-proc.sh functions. Remove the /etc/sysconfig
      stuff entirely (it was not being used anyway).
    - fix for #795934 : after rpm -U, pcp is chkconfig off. It turned
      out that an upgrade executes the %post _and_ the %preun scripts,
      which resulted in pcp being chkconfig'd on then off again.
    - fix pmie rc scripts so they work, are chkconfig friendly,
      and cope with _and_ without pmieconf (which is in pcp-pro).
      Also install /var/pcp/config/pmie/config.default as a simple example
      to monitor the load average and report to syslog. The pmie
      daemon is chkconfig off by default.
    - default run levels for pmcd and pmie (daemon) are now 2345, for SUSE
    - reconcile troff and groff differences in man page sources
    - fix for bug #797049 use strftime(%z) to determine timezone offsets
      w.r.t. daylight savings
    - portability surgery on src/libpcp_trace, and add new pmtracecounter()

src/query/kmquery.cpp
/*
 * Copyright (c) 2007, Aconex.  All Rights Reserved.
 * 
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 */
#include "kmquery.h"

#define max(a,b) ((a)>(b)?(a):(b))
#define DEFAULT_EDIT_WIDTH	420	/* in units of pixels */
#define DEFAULT_EDIT_HEIGHT	160	/* only for usesliderflag */

enum icontypes {
    INFO_ICON,
    ERROR_ICON,
    QUESTION_ICON,
    WARNING_ICON,
    ARCHIVE_ICON,
    HOST_ICON,
} iconic;

static const char *title = "Query";
static int timeout;
static int *statusi;
static const char **buttons;
static int buttoncount;
static const char *defaultbutton;
static char **messages;
static int messagecount;

static void nomem()
{
    fputs("Insufficient memory\n", stderr);
    exit(1);
}

int KmQuery::setTimeout(char *string)
{
    char *endnum;
    timeout = (int)strtol(string, &endnum, 10);
    if (*endnum != '\0' || timeout <= 0)
	return -1;
    return 0;
}

void KmQuery::setTitle(char *heading)
{
    title = heading;
}

int KmQuery::messageCount()
{
    return messagecount;
}

int KmQuery::buttonCount()
{
    return buttoncount;
}

int KmQuery::setIcontype(char *string)
{
    if (strcmp(string, "info") == 0)
	iconic = INFO_ICON;
    else if (strcmp(string, "error") == 0 ||
	strcmp(string, "action") == 0 ||
	strcmp(string, "critical") == 0)
	iconic = ERROR_ICON;
    else if (strcmp(string, "question") == 0)
	iconic = QUESTION_ICON;
    else if (strcmp(string, "warning") == 0)
	iconic = WARNING_ICON;
    else if (strcmp(string, "archive") == 0)
	iconic = ARCHIVE_ICON;
    else if (strcmp(string, "host") == 0)
	iconic = HOST_ICON;
    else
	return -1;
    return 0;
}

void KmQuery::addMessage(char *string)
{
    messages = (char **)realloc(messages, (messagecount+1) * sizeof(char *));
    if (!messages)
	nomem();
    messages[messagecount++] = string;
}

void KmQuery::addButton(const char *string, bool iamdefault, int status)
{
    buttons = (const char **)realloc(buttons, (buttoncount+1) * sizeof(char *));
    statusi = (int *)realloc(statusi, (buttoncount+1) * sizeof(int));
    if (!buttons)
	nomem();
    if (iamdefault)
	defaultbutton = string;
    statusi[buttoncount] = status;
    buttons[buttoncount++] = string;
}

void KmQuery::addButtons(char *string) // comma-separated label:exitcode string
{
    char *n;
    QString pairs(string);
    static int next = 100;

    QStringList list = pairs.split(",");
    for (QStringList::Iterator it = list.begin(); it != list.end(); ++it) {
	QString name = (*it).section(":", 0, 0);
	QString code = (*it).section(":", 1, 1);
	if (!name.isEmpty()) {
	    int sts = code.isEmpty() ? ++next : code.toInt();
	    if ((n = strdup(name.toAscii().data())) == NULL)
		nomem();
	    addButton(n, FALSE, sts);
	}
    }
}

void KmQuery::setDefaultButton(char *string)
{
    for (int i = 0; i < buttoncount; i++)
	if (strcmp(buttons[i], string) == 0)
	    defaultbutton = buttons[i];
}

void KmQuery::buttonClicked()
{
    done(my.status);
}

void KmQuery::timerEvent(QTimerEvent *)
{
    done(1);
}

// Currently we set default edit size to hardcoded values, until
// better ways are found to interact with any geometry requests.
// Note: the +4 pixels for height ensure the auto-scroll does not
// kick in, seems to be required.

KmQuery::KmQuery(bool inputflag, bool printflag, bool noframeflag,
		 bool nosliderflag, bool usesliderflag, bool exclusiveflag)
    : QDialog()
{
    QHBoxLayout *hboxLayout;
    QVBoxLayout *vboxLayout;
    QSpacerItem *spacerItem;
    QSpacerItem *spacerItem1;
    QVBoxLayout *vboxLayout1;
    QHBoxLayout *hboxLayout1;
    QSpacerItem *spacerItem2;

    QString filename;
    if (iconic == HOST_ICON)
	filename = tr(":dialog-host.png");
    else if (iconic == ERROR_ICON)
	filename = tr(":dialog-error.png");
    else if (iconic == WARNING_ICON)
	filename = tr(":dialog-warning.png");
    else if (iconic == ARCHIVE_ICON)
	filename = tr(":dialog-archive.png");
    else if (iconic == QUESTION_ICON)
	filename = tr(":dialog-question.png");
    else // (iconic == INFO_ICON)
	filename = tr(":dialog-information.png");

    QIcon	icon(filename);
    QPixmap	pixmap(filename);
    setWindowIcon(icon);
    setWindowTitle(tr(title));

    QGridLayout *gridLayout = new QGridLayout(this);
    gridLayout->setSpacing(6);
    gridLayout->setMargin(9);
    hboxLayout = new QHBoxLayout();
    hboxLayout->setSpacing(6);
    hboxLayout->setMargin(0);
    vboxLayout = new QVBoxLayout();
    vboxLayout->setSpacing(6);
    vboxLayout->setMargin(0);
    spacerItem = new QSpacerItem(20, 2, QSizePolicy::Minimum,
					QSizePolicy::Expanding);

    vboxLayout->addItem(spacerItem);

    QLabel *imageLabel = new QLabel(this);
    imageLabel->setPixmap(pixmap);

    vboxLayout->addWidget(imageLabel);

    spacerItem1 = new QSpacerItem(20, 20, QSizePolicy::Minimum,
					  QSizePolicy::Expanding);

    vboxLayout->addItem(spacerItem1);
    hboxLayout->addLayout(vboxLayout);
    vboxLayout1 = new QVBoxLayout();
    vboxLayout1->setSpacing(6);
    vboxLayout1->setMargin(0);

    int height;
    int width = DEFAULT_EDIT_WIDTH; 

    QLineEdit* lineEdit = NULL;
    QTextEdit* textEdit = NULL;
    if (inputflag && messagecount <= 1) {
	lineEdit = new QLineEdit(this);
	if (messagecount == 1)
	    lineEdit->setText(tr(messages[0]));
	height = lineEdit->font().pointSize() + 4;
	if (height < 0)
	    height = lineEdit->font().pixelSize() + 4;
	if (height < 0)
	    height = lineEdit->heightForWidth(width) + 4;
	lineEdit->setSizePolicy(QSizePolicy::MinimumExpanding,
				QSizePolicy::Fixed);
	lineEdit->setMinimumSize(QSize(width, height));
	lineEdit->setGeometry(QRect(0, 0, width, height));
	vboxLayout1->addWidget(lineEdit);
    }
    else {
	textEdit = new QTextEdit(this);
	textEdit->setLineWrapMode(QTextEdit::FixedColumnWidth);
	textEdit->setLineWrapColumnOrWidth(80);
	textEdit->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
	if (nosliderflag)
	    textEdit->setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
	else if (usesliderflag)
	    textEdit->setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOn);
	else
	    textEdit->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
	for (int m = 0; m < messagecount; m++)
	    textEdit->append(tr(messages[m]));
	if (inputflag)
	    textEdit->setReadOnly(FALSE);
	else {
	    textEdit->setLineWidth(1);
	    textEdit->setFrameStyle(noframeflag ? QFrame::NoFrame :
				    QFrame::Box | QFrame::Sunken);
	    textEdit->setReadOnly(TRUE);
	}
	if (usesliderflag)
	    height = DEFAULT_EDIT_HEIGHT;
	else {
	    height = textEdit->font().pointSize() + 4;
	    if (height < 0)
		height = textEdit->font().pixelSize() + 4;
	    if (height < 0)
	        height = textEdit->heightForWidth(width) + 4;
	    height *= messagecount;
	}
	textEdit->setMinimumSize(QSize(width, height));
	textEdit->setSizePolicy(QSizePolicy::MinimumExpanding,
				QSizePolicy::MinimumExpanding);
	vboxLayout1->addWidget(textEdit);
    }

    hboxLayout1 = new QHBoxLayout();
    hboxLayout1->setSpacing(6);
    hboxLayout1->setMargin(0);
    spacerItem2 = new QSpacerItem(40, 20, QSizePolicy::Expanding,
					  QSizePolicy::Minimum);
    hboxLayout1->addItem(spacerItem2);

    for (int i = 0; i < buttoncount; i++) {
	QueryButton *button = new QueryButton(printflag, this);
	button->setMinimumSize(QSize(72, 32));
	button->setDefault(buttons[i] == defaultbutton);
	button->setQuery(this);
	button->setText(tr(buttons[i]));
	button->setStatus(statusi[i]);
	if (inputflag && buttons[i] == defaultbutton) {
	    if (textEdit) 
		button->setEditor(textEdit);
	    else if (lineEdit) {
		button->setEditor(lineEdit);
		if (buttons[i] == defaultbutton)
		    connect(lineEdit, SIGNAL(returnPressed()),
			    button, SLOT(click()));
	    }
	}
	connect(button, SIGNAL(clicked()), this, SLOT(buttonClicked()));
	hboxLayout1->addWidget(button);
    }

    vboxLayout1->addLayout(hboxLayout1);
    hboxLayout->addLayout(vboxLayout1);
    gridLayout->addLayout(hboxLayout, 0, 0, 1, 1);
    gridLayout->setSizeConstraint(QLayout::SetFixedSize);

    if (inputflag && messagecount <= 1)
	resize(QSize(320, 83));
    else
	resize(QSize(320, 132));

    if (timeout)
	startTimer(timeout * 1000);

    if (exclusiveflag)
	setWindowModality(Qt::WindowModal);
}
